Updating reqwest to 0.11.22 causes macos build failure, requires SystemConfiguration

In some dev work, I updated `reqwest` from 0.11.20 to 0.11.22 and ran into [failures](https://github.com/subconsciousnetwork/noosphere/actions/runs/6566865808/job/17840254182?pr=549) in a x86_64 macos build:

```
Undefined symbols for architecture x86_64:
  "_SCDynamicStoreCopyProxies", referenced from:
      __ZN20system_configuration13dynamic_store14SCDynamicStore11get_proxies17ha9cee4367b0c1fcbE in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCDynamicStoreCreateWithOptions", referenced from:
      __ZN20system_configuration13dynamic_store14SCDynamicStore6create17he63aeeeb3840fdc4E in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCNetworkInterfaceCopyAll", referenced from:
      __ZN20system_configuration21network_configuration14get_interfaces17ha41c0e3fd9b2cb6eE in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCNetworkInterfaceGetBSDName", referenced from:
      __ZN20system_configuration21network_configuration18SCNetworkInterface8bsd_name17h45a02c56c7f559b2E in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCNetworkInterfaceGetInterfaceType", referenced from:
      __ZN20system_configuration21network_configuration18SCNetworkInterface21interface_type_string17hd31e474b334507ecE in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCNetworkInterfaceGetLocalizedDisplayName", referenced from:
      __ZN20system_configuration21network_configuration18SCNetworkInterface12display_name17h94d4d8074b27dd32E in libnoosphere.a(system_configuration-46c6672ca6db6ca9.system_configuration.b48ec42a4ca5d16c-cgu.0.rcgu.o)
  "_SCNetworkReachabilityCreateWithAddress", referenced from:
```

In https://github.com/seanmonstar/reqwest/pull/1955, [system-configuration](https://github.com/mullvad/system-configuration-rs) was added to reqwest, requiring linking to the SystemConfiguration framework. The transitive dependency seems to link the framework in its [build script](https://github.com/mullvad/system-configuration-rs/blob/main/system-configuration-sys/build.rs) at a glance, but we hit undefined symbol errors. Need to evaluate if this is an intended required change, and if so, learn more about if we want to require [SystemConfiguration.framework](https://developer.apple.com/documentation/systemconfiguration) on macos.

---

Upstream bug in system-configuration: https://github.com/mullvad/system-configuration-rs/issues/41
Patch for iOS, to investigate: https://github.com/mullvad/system-configuration-rs/pull/42/files

---

Filed upstream Reqwest bug: https://github.com/seanmonstar/reqwest/issues/2006

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -60,7 +60,7 @@ libipld-cbor = { version = "0.16" }
 libipld-json = { version = "0.16" }
 pathdiff = { version = "0.2.1" }
 rand = { version = "0.8" }
-reqwest = { version = "=0.11.20", default-features = false, features = ["json", "rustls-tls", "stream"] }
+reqwest = { version = "0.12", default-features = false, features = ["json", "rustls-tls", "stream"] }
 sentry-tracing = { version = "0.31.8" }
 serde = { version = "^1" }
 serde_json = { version = "^1" }
@@ -88,7 +88,7 @@ void = { version = "1" }
 wasm-bindgen = { version = "^0.2" }
 wasm-bindgen-test = { version = "^0.3" }
 wasm-bindgen-futures = { version = "^0.4" }
-wasm-streams = { version = "0.3.0" }
+wasm-streams = { version = "0.4" }
 web-sys = { version = "0.3" }
 
 [profile.release]

--- a/rust/noosphere-cli/Cargo.toml
+++ b/rust/noosphere-cli/Cargo.toml
@@ -28,7 +28,7 @@ observability = ["noosphere-gateway/observability"]
 vergen = { workspace = true }
 
 [target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
-reqwest = { version = "~0.11", default-features = false, features = ["json", "rustls-tls", "stream"] }
+reqwest = { workspace = true, default-features = false, features = ["json", "rustls-tls", "stream"] }
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 tempfile = { workspace = true }

--- a/rust/noosphere-core/src/api/client.rs
+++ b/rust/noosphere-core/src/api/client.rs
@@ -1,7 +1,7 @@
-use std::str::FromStr;
-
 use crate::{
-    api::{route::RouteUrl, v0alpha1, v0alpha2, StatusCode},
+    api::{route::RouteUrl, v0alpha1, v0alpha2},
+    authority::{generate_capability, Author, SphereAbility, SphereReference},
+    data::{Link, MemoIpld},
     error::NoosphereError,
     stream::{from_car_stream, memo_history_stream, put_block_stream, to_car_stream},
 };
@@ -9,14 +9,10 @@ use anyhow::{anyhow, Result};
 use async_stream::try_stream;
 use bytes::Bytes;
 use cid::Cid;
+use http::StatusCode;
 use iroh_car::CarReader;
 use libipld_cbor::DagCborCodec;
 use noosphere_common::{ConditionalSend, ConditionalSync, UnsharedStream};
-
-use crate::{
-    authority::{generate_capability, Author, SphereAbility, SphereReference},
-    data::{Link, MemoIpld},
-};
 use noosphere_storage::{block_deserialize, block_serialize, BlockStore};
 use noosphere_ucan::{
     builder::UcanBuilder,
@@ -26,6 +22,7 @@ use noosphere_ucan::{
     ucan::Ucan,
 };
 use reqwest::header::HeaderMap;
+use std::str::FromStr;
 use tokio_stream::{Stream, StreamExt};
 use tokio_util::io::StreamReader;
 use url::Url;
@@ -93,7 +90,7 @@ where
             client.get(url).send().await?
         };
 
-        match translate_status_code(did_response.status())? {
+        match did_response.status() {
             StatusCode::OK => (),
             _ => return Err(anyhow!("Unable to look up gateway identity")),
         };
@@ -475,7 +472,7 @@ where
                 v0alpha2::PushError::BrokenUpstream
             })?;
 
-        let status = translate_status_code(response.status())?;
+        let status = response.status();
         trace!("Checking response ({})...", status);
 
         if status == StatusCode::CONFLICT {
@@ -530,23 +527,4 @@ where
 
         Ok(push_response)
     }
-}
-
-/// Both `reqwest` and `axum` re-export `StatusCode` from the `http` crate.
-///
-/// We're stuck on reqwest@0.11.20 [1] that uses an older version
-/// of `http::StatusCode`, whereas axum >= 0.7 uses the 1.0 release
-/// of several HTTP libraries (`http`, `http-body`, `hyper`) [2], which
-/// we'd like to use as our canonical representation.
-///
-/// This utility converts between the old `reqwest::StatusCode` to the
-/// >=1.0 implementation. Notably, we do not pull in all of `axum`
-/// into the `noosphere-core` crate, only the common underlying
-/// crate `http@1.0.0` (or greater).
-///
-/// [1] https://github.com/subconsciousnetwork/noosphere/issues/686
-/// [2] https://github.com/tokio-rs/axum/blob/5b6204168a676497d2f4188af603546d9ebfe20a/axum/CHANGELOG.md#070-27-november-2023
-fn translate_status_code(reqwest_code: reqwest::StatusCode) -> Result<StatusCode> {
-    let code: u16 = reqwest_code.into();
-    Ok(code.try_into()?)
 }

--- a/rust/noosphere-core/src/api/mod.rs
+++ b/rust/noosphere-core/src/api/mod.rs
@@ -11,7 +11,3 @@ pub mod v0alpha2;
 
 pub use client::*;
 pub use data::*;
-
-// Re-export `http::StatusCode` here as our preferred `StatusCode` instance,
-// disambiguating from other crate's implementations.
-pub(crate) use http::StatusCode;

--- a/rust/noosphere-core/src/api/v0alpha1/data.rs
+++ b/rust/noosphere-core/src/api/v0alpha1/data.rs
@@ -1,15 +1,11 @@
-use std::fmt::Display;
-
-use crate::api::{
-    data::{empty_string_as_none, AsQuery},
-    StatusCode,
-};
 use crate::{
+    api::data::{empty_string_as_none, AsQuery},
     authority::{generate_capability, SphereAbility, SPHERE_SEMANTICS},
     data::{Bundle, Did, Jwt, Link, MemoIpld},
 };
 use anyhow::{anyhow, Result};
 use cid::Cid;
+use http::StatusCode;
 use noosphere_storage::{base64_decode, base64_encode};
 use noosphere_ucan::{
     chain::ProofChain,
@@ -18,6 +14,7 @@ use noosphere_ucan::{
     Ucan,
 };
 use serde::{Deserialize, Serialize};
+use std::fmt::Display;
 use thiserror::Error;
 
 /// The query parameters expected for the "replicate" API route.

--- a/rust/noosphere-core/src/api/v0alpha2/data.rs
+++ b/rust/noosphere-core/src/api/v0alpha2/data.rs
@@ -1,7 +1,5 @@
-use crate::{
-    api::StatusCode,
-    data::{Did, Jwt, Link, MemoIpld},
-};
+use crate::data::{Did, Jwt, Link, MemoIpld};
+use http::StatusCode;
 use serde::{Deserialize, Serialize};
 use thiserror::Error;
 

--- a/rust/noosphere-ipfs/Cargo.toml
+++ b/rust/noosphere-ipfs/Cargo.toml
@@ -31,7 +31,7 @@ async-stream = { workspace = true }
 libipld-core = { workspace = true }
 libipld-cbor = { workspace = true }
 cid = { workspace = true }
-reqwest = { version = "~0.11", default-features = false, features = ["json", "rustls-tls", "stream"] }
+reqwest = { workspace = true, default-features = false, features = ["json", "rustls-tls", "stream"] }
 serde = { workspace = true }
 serde_json = { workspace = true }
 tokio = { workspace = true, features = ["io-util"] }

--- a/rust/noosphere-ns/Cargo.toml
+++ b/rust/noosphere-ns/Cargo.toml
@@ -51,7 +51,7 @@ toml = { version = "~0.8", optional = true }
 # noosphere_ns::server
 axum = { workspace = true, features = ["json", "macros"], optional = true }
 axum-tracing-opentelemetry = { workspace = true, optional = true }
-reqwest = { version = "~0.11", default-features = false, features = ["json", "rustls-tls"], optional = true }
+reqwest = { workspace = true, default-features = false, features = ["json", "rustls-tls"], optional = true }
 tower-http = { workspace = true, features = ["trace"], optional = true }
 url = { version = "^2", features = [ "serde" ], optional = true }
 

